lib/arel/visitors/oracle_common.rb

```diff
@@ -4,14 +4,61 @@ module Arel # :nodoc: all
   module Visitors
     module OracleCommon
       private
-        def visit_Arel_Nodes_Matches(o, collector)
-          if !o.case_sensitive && o.left && o.right
-            o.left = Arel::Nodes::NamedFunction.new("UPPER", [o.left])
-            o.right = Arel::Nodes::NamedFunction.new("UPPER", [o.right])
+
+      # Oracle can't compare CLOB columns with standard SQL operators for comparison.
+      # We need to replace standard equality for text/binary columns to use DBMS_LOB.COMPARE function.
+      # Fixes ORA-00932: inconsistent datatypes: expected - got CLOB
+      def visit_Arel_Nodes_Equality(o, collector)
+        case (left = o.left)
+        when Arel::Attributes::Attribute
+          column = column_for(left)
+
+          case column.type
+          when :text, :binary
+            # https://docs.oracle.com/cd/B19306_01/appdev.102/b14258/d_lob.htm#i1016668
+            # returns 0 when the comparison succeeds
+            comparator = Arel::Nodes::NamedFunction.new('DBMS_LOB.COMPARE', [left, o.right])
+            collector = visit comparator, collector
+            collector << ' = 0'
+            collector
+          else
+            super
           end
+        else
+          super
+        end
+      end
 
-          super o, collector
+      def visit_Arel_Nodes_Matches(o, collector)
+        if !o.case_sensitive && o.left && o.right
+          o.left = Arel::Nodes::NamedFunction.new("UPPER", [o.left])
+          o.right = Arel::Nodes::NamedFunction.new("UPPER", [o.right])
         end
+
+        super o, collector
+      end
+
+      def column_for attr
+        return unless attr
+        name = attr.name.to_s
+        table = attr.relation.table_name
+
+        return nil unless table_exists? table
+
+        column_cache(table)[name]
+      end
+
+      def column_cache(table)
+        schema_cache.columns_hash(table)
+      end
+
+      def schema_cache
+        @connection.schema_cache
+      end
+
+      def table_exists? name
+        schema_cache.data_source_exists?(name)
+      end
```

Comment:
These are copied off [arel 7.1.4](https://github.com/rails/arel/blob/v7.1.4/lib/arel/visitors/to_sql.rb). These methods were removed from latter arel releases but are needed for `#visit_Arel_Nodes_Equality` to function.